test/tests/functional/pbs_hook_execjob_end.py

```diff
@@ -0,0 +1,162 @@
+            date_time1, '%m/%d/%Y %H:%M:%S')))
+        (_, str1) = self.mom.log_match("executed exechost_periodic hook",
+                                       starttime=epoch1, n=100,
+                                       max_attempts=10, interval=2)
+        date_time2 = str1.split(";")[0]
+        epoch2 = int(time.mktime(time.strptime(
+            date_time2, '%m/%d/%Y %H:%M:%S')))
+        (_, str1) = self.mom.log_match(
+            "Job;%s;execjob_end hook ended" %
+            jid, starttime=epoch2, n=100,
+            max_attempts=10, interval=2)
+        date_time3 = str1.split(";")[0]
+        self.logger.info(
+            "execjob_end hook executed at: %s,"
+            "exechost_periodic at: %s and execjob_end hook ended at: %s" %
+            (date_time1, date_time2, date_time3))
+
+    def test_execjob_end_hook_order_and_reject(self):
+        """
+        Test with multiple execjob_end hooks having different order
+        with one of the hooks rejecting the job.
+        """
+        hook_name1 = "execjob_end_logmsg1"
+        hook_body = ("import pbs\n"
+                     "e = pbs.event()\n"
+                     "pbs.logjobmsg(e.job.id, \
+                                  'executed %s hook' % e.hook_name)\n"
+                     "e.accept()\n")
+        attr = {'event': 'execjob_end', 'order': '1', 'enabled': 'True'}
+        self.server.create_import_hook(hook_name1, attr, hook_body)
+        hook_name = "execjob_end_logmsg2"
+        hook_body1 = ("import pbs\n"
+                      "e = pbs.event()\n"
+                      "pbs.logjobmsg(e.job.id, 'executed execjob_end hook')\n"
+                      "e.reject('Job is rejected')\n")
+        attr = {'event': 'execjob_end', 'order': '2', 'enabled': 'True'}
+        self.server.create_import_hook(hook_name, attr, hook_body1)
+        hook_name2 = "execjob_end_logmsg3"
+        attr = {'event': 'execjob_end', 'order': '170', 'enabled': 'True'}
+        self.server.create_import_hook(hook_name2, attr, hook_body)
+        j = Job(TEST_USER)
+        j.set_sleep_time(1)
+        jid = self.server.submit(j)
+        self.mom.log_match("Job;%s;executed %s hook" % (jid, hook_name1),
+                           n=100, max_attempts=10, interval=2)
+        self.mom.log_match("Job;%s;Job is rejected" % jid,
+                           n=100, max_attempts=10, interval=2)
+        self.mom.log_match("Job;%s;executed %s hook" % (jid, hook_name2),
+                           n=100, max_attempts=10, interval=2, existence=False)
+
+    def test_execjob_end_multi_job(self):
+        """
+        Test a execjob_end hook for non-blocking of mom with mutiple jobs
+        """
+        hook_name = "execjob_end_logmsg4"
+        self.server.create_import_hook(hook_name, self.attr, self.hook_body)
+        j = Job(TEST_USER)
+        j.set_sleep_time(1)
+        jid1 = self.server.submit(j)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid1)
```

Comment:
This will cause a race condition since 1s is too short. remove this check.

Reply:
This check is required since we want job 2 to be submitted after job1 has started.  I'll increase the sleep time of job1 to 3s Instead to avoid race condition.

Reply:
@saimanikant 3s is also not enough. Either you put 10s or add a sleep of 2s between job submissions or make sure job1 is finished before submitting job2.